### DISCUSSION TITLE
Add GET support for astro:actions

### DISCUSSION THREAD

**Discussion by WesSouza** [withastro/roadmap#1064] [Proposal]
# Summary

Currently, Astro Actions can be used to send data and perform actions on the server via POST requests.

However, users might want to get data from the server while leveraging browser and server caching, which requires a GET request.

# Background & Motivation

I am working on a project and wanted to use astro:actions for its tRPC-like ergonomics. My project contains several framework components which then obtain data from the server utilizing actions.

Those data requests could be cached so concurrent users don't have to reexecute the action, but given all actions are POST, they must reevaluate.

A use case I immediately have is to allow access to collection content on my framework components without sending the entire collection as a prop.

# Goals

- Update `defineAction` to indicate the use of the GET method (such as with a new property `method: 'get'`;
- Add a GET handler on [packages/astro/src/actions/runtime/route.ts](https://github.com/withastro/astro/blob/main/packages/astro/src/actions/runtime/route.ts);
- Allow either object-to-query-string or serialize it using `JSON.stringify` and send as the `input` query parameter.

# Note

It must be documented that there would be a size limitation on input parameters. However, I do not see it as an impediment given GET API inputs are usually related to filtering and therefore succinct.

# Example

```ts
export const server = {
  getCollection: defineAction({
    method: 'get',
    input: z.object({ type: z.enum(['blog', 'documents']) }),
    handler: async (input) => {
      const content = await getCollection(input.type);
      return content.map((item) => ({
        id: item.id,
        collection: item.collection,
        data: item.data,
        slug: item.slug,
      }));
    },
  })
}
```

**Comment by bholmesdev** [comment]
Oh, this is a nice suggestion! I wonder if we can expand our current `accept` property to represent this. Perhaps an `accept: 'search'` alongside `accept: 'form'` and `accept: 'json'`? This follows our pattern of using `accept` to determine how payloads are parsed.

That said, I can see `method: 'get'` more directly answering the need people have: I want GET so I can cache.

**Reply by WesSouza** [reply]
I do think it's a different perspective, as you could have both `method: 'get'` and `accept: 'form'` if you use a `<form>` without `method`, which would generate a GET request.

Basically `method`, or another keyword, would convey the transport method for the request, and `accept` the encoding.

To me the four options would be:

| Method | Accept | Coding Method |
| ------ | ------ | -------------- |
| `get`  | `form` | Identical to [the current approach](https://docs.astro.build/en/reference/modules/astro-actions/#use-with-accept-form), except obviously for `File`. |
| `get`  | `json` | The query string would serialize the value into the `input` parameter using `JSON.stringify`, and then `JSON.parse` + Zod's `parse`. |
| `post` | `form` | No change. |
| `post` | `json` | No change. |

We could also follow [zod-search-params](https://github.com/theboxer/zod-search-params) or [can-param](https://canjs.com/doc/can-param.html) and be more like PHP which allows any sort of object to be represented as a query parameter.

In these cases:

## `method: 'get', accept: 'form'`

```html
<form action={actions.example.url}>
  <input name="test" value="123">
  <button>Send</button>
</form>
```

`actions.example.url` could map to the endpoint so it would be natively usable by a `<form>` element.

Request:
```
GET /_actions/example?test=123
```

```js
export const server = {
  example: defineAction({
    method: 'get',
    accept: 'form',
    input: z.object({ test: z.string() }),
    handler: async (input) => {
      // ...
    },
  })
}
```

## `method: 'get', accept: 'json'`


Client:

```js
actions.example({ test: '123' })
```

Request:
```
GET /_actions/example?input=%7B%22test%22%3A%22123%22%7D
```

Server:

```js
export const server = {
  example: defineAction({
    method: 'get',
    accept: 'json',
    input: z.object({ test: z.string() }),
    handler: async (input) => {
      // ...
    },
  })
}
```

**Reply by WesSouza** [reply]
BTW I looked at the code and see how a `getSearchParamsServerHandler` could be implemented, and convey the meaning the action would take place in a GET request.

**Reply by bholmesdev** [reply]
Ah ha, if we want to serialize form and json payloads separately, I could definitely see that working! I was imagining you pass a `URLSearchParams()` object from the client, but I think an object -> search params serializer would be more intuitive. I'd be on board with this

**Reply by lilnasy** [reply]
I ended up going with an approach similar to method: get when making [astro-typed-api](https://github.com/lilnasy/gratelets/tree/main/packages/typed-api). The [de/serializer](https://github.com/lilnasy/gratelets/blob/main/packages/typed-api/runtime/param-codec.ts) is pretty simple. 

Another option is using a request header, and adding that header name to the response's [`Vary`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Vary) header list. And, [trpc](https://trpc.io/docs/rpc) allows caching by json encoding followed by uri encoding the arguments into a search parameter.

**Reply by TranscendentalRide** [reply]
Want to give my 2-cents here.

Support GET would be a huge win. Specifically because GET supports conditional requests. This where client sends and `IF-NONE-MATCH` header, a hash received from the server in a prior call. The server still process the request; but only sends a payload if results doesn't match the ETAG value. If there is a match, the server sends a 304 (Not Modified) instead of a 200. Saving bandwidth and time.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Status/304

Tanstack supports this:

https://tanstack.com/start/latest/docs/framework/react/server-functions

```ts

export const getData = createServerFn({
  method: 'GET', // HTTP method to use
  response: 'data', // Response handling mode
}).handler(async () => {
  // Function implementation
})
```

**Comment by Eptagone** [comment]
I need this too. Also, it would be great if GET was the default when no input arguments are needed.

**Comment by ausminternet** [comment]
Any updates on this?

**Reply by bholmesdev** [reply]
To be transparent: I've stepped away from Astro contributions recently, and I don't think Wes is available / interested in leading the feature. I'll bring this to our maintainers to see if it's roadmap worthy in the near term.

Otherwise, this is a call to anyone following this issue to lead the implementation! I'm happy to assist in PR reviews or questions.

**Comment by Ely-S** [comment]
I would love to see this feature, and am able and willing to contribute it!

There is some discussion here on how it should be designed. I'm open to the different approaches, as well. Do any of the maintainers have strong feelings on what the implementation should look like?

**Reply by bholmesdev** [reply]
Amazing! After seeing a couple options, I think adding a `method` field would make the most sense. The alternative would be to add `search-params` as an `accept` option, but I don't think it's intuitive to switch from POST to GET by doing so.

If we go with `method: 'get'`, I'd be careful about:
1. How we handle the `input` schema. It should only allow strings, numbers, and booleans, and arrays of each of these
2. How we update the valid types for `accept`. `accept: json` and `accept: form` don't make sense for a GET request. We can either disallow `accept` when a `method` is present, or we can allow `accept: 'search-params'` only. Not opposed to either

**Reply by bholmesdev** [reply]
Another alternative may be a `defineGetAction()` that papers over this type problem. But it wouldn't be consistent with how the rest of the API is designed using object properties.

If you have further ideas, I'm open to discussing!

**Reply by Ely-S** [reply]
Regarding the input handling, here are my ideas:

1. Keep `accept json` to allow nested objects. We would JSON encode the request body that would be sent for POST, and send it as a search parameter, such as:

`?_astroActionPayload=${encodeURIComponent(JSON.stringify(body))}`

And then a small amount of backend logic would be required to parse it. I would make the serialization sort the json by key, so that it is more cacheable.

2. We could also have `accept: search-params`, which could only accept data as simple query parameters, with only strings, numbers, booleans, and arrays.

If we choose to limit the input types for GET requests, then we should require `accept: search-params` so that it is clear nested objects are not allowed.

**Reply by bholmesdev** [reply]
@Ely-S Okay yeah, I can get behind stringifying the payload to JSON by default. In fact, I'd prefer that to introducing `accept: search-params` (unless there's a clear user need to parse those search params from Astro.url. Don't think there is). Sorting for cacheability is a good call as well.

If you want, you can [make a stage 2 issue with that proposed solution](https://github.com/withastro/roadmap?tab=readme-ov-file#stage-2-accepted-proposal) to keep the ball rolling.

**Reply by Ely-S** [reply]
Hi, I followed up and created a stage 2 issue #1345
